### PR TITLE
Amr Yehia week3 JavaScript

### PR DESCRIPTION
**Pull request by Amr-Yehia**

### CODE REVIEW

**Comment by Amr-Yehia** [comment]
completed the homework

File: 1-JavaScript/Week3/homework/ex1-giveCompliment.js [hunk 1]
```diff
@@ -18,13 +18,29 @@ Full description at: https://github.com/HackYourFuture/Homework/tree/main/1-Java
...
+    'fabulous',
+    'helpful',
+    'wonderful',
+    'nice',
+    'lovely',
+    'fantastic',
+    'honest',
+  ];
+  const compliment =
+    compliments[Math.floor(Math.random() * compliments.length)];
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week3/homework/ex1-giveCompliment.js:36]: Excellent 🎉

File: 1-JavaScript/Week3/homework/ex2-dogYears.js [hunk 2]
```diff
@@ -16,8 +16,12 @@ calculate it!
    ages.
 -----------------------------------------------------------------------------*/
 
-function calculateDogAge(/* TODO parameter(s) go here */) {
+function calculateDogAge(fictional) {
   // TODO complete this function
+  const age = fictional * 7;
+  return `Your doggie is ${age} years old in dog years!`;
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week3/homework/ex2-dogYears.js:22]: Perfect!

File: 1-JavaScript/Week3/homework/ex3-tellFortune.js [hunk 3]
```diff
@@ -22,40 +17,62 @@ Why pay a fortune teller when you can just program your fortune yourself?
...
 randomly select array elements four times inside the `tellFortune` function 
 body, this code is now written once only in a separated function.
 -----------------------------------------------------------------------------*/
 
 // This function should take an array as its parameter and return
 // a randomly selected element as its return value.
-function selectRandomly(/* TODO parameter(s) go here */) {
+function selectRandomly(arrays) {
   // TODO complete this function
+  const random = Math.floor(Math.random() * 5);
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week3/homework/ex3-tellFortune.js:31]: This works! Ask yourself if it always works. Eg: what would your function do if you would call:
```javascript
selectRandomly(['apple', 'banana', 'strawberry']);
```

Can you think of a way to make this function work for arrays of different lengths? Hint: take a look at your own `giveCompliments` implementation :-)

File: 1-JavaScript/Week3/homework/ex4-shoppingCart.js [hunk 4]
```diff
@@ -1,32 +1,36 @@
...
 
 // ! Function to be tested
-function addToShoppingCart(/* parameters go here */) {
+function addToShoppingCart(item) {
   // TODO complete this function
-}
 
+  let cart = shoppingCart.length;
+  shoppingCart.push(item);
+  const newShoppingCart =  shoppingCart.filter(e =>  e);
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week3/homework/ex4-shoppingCart.js:26]: We slacked about this and I know why you introduced it. But if I look at your current code, is it still necessary? Just try it out if you can't predict what will happen and see if you can understand what happens.

File: 1-JavaScript/Week3/homework/ex4-shoppingCart.js [hunk 5]
```diff
@@ -1,32 +1,36 @@
...
+  if(cart <= 2){
+    return `You bought ${newShoppingCart.join(', ')}!`;
+  }else if(cart > 2){
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week3/homework/ex4-shoppingCart.js:29]: You implemented the two conditions correctly. Something to consider: could you leave out the second if statement? So what would happen if you would write:
```javascript
 if(cart <= 2){
     return `You bought ${newShoppingCart.join(', ')}!`;
   } else {
...
```

File: 1-JavaScript/Week3/homework/ex5-shoppingCartPure.js [hunk 6]
```diff
@@ -15,9 +15,17 @@ it pure. Do the following:
...
 ------------------------------------------------------------------------------*/
+
+ 
 // ! Function under test
-function addToShoppingCart(/* TODO parameter(s) go here */) {
+function addToShoppingCart(cart, item) {
   // TODO complete this function
+  const newCart = [...cart, item]
+  if(newCart.length > 3){
+      return newCart.slice(newCart.length - 3);
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week3/homework/ex5-shoppingCartPure.js:25]: you definately made it work! Did you know that slice can also take a negative number that counts from the end of the array? 
```javascript
const animals = ['ant', 'bison', 'camel', 'duck', 'elephant'];

console.log(animals.slice(-2));
// expected output: Array ["duck", "elephant"]
```

File: 1-JavaScript/Week3/homework/ex6-totalCost.js [hunk 7]
```diff
@@ -22,21 +22,41 @@ instead!
...
 
-function calculateTotalPrice(/* TODO parameter(s) go here */) {
+function calculateTotalPrice(part) {
   // TODO replace this comment with your code
+  
+  let sum = 0;
+  for(const price in cartForParty ){
+      sum += cartForParty[price]
+  }
+  return `Total: € ${sum}`;
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week3/homework/ex6-totalCost.js:39]: Awesome :-)

File: 1-JavaScript/Week3/homework/ex7-mindPrivacy.js [hunk 8]
```diff
@@ -30,8 +30,13 @@ const employeeRecords = [
 ];
 
 // ! Function under test
-function filterPrivateData(/* TODO parameter(s) go here */) {
+function filterPrivateData(record/* TODO parameter(s) go here */) {
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week3/homework/ex7-mindPrivacy.js:33]: I can see you struggled a bit here, so I'll try and help you a little without giving the answers directly. Hope that is ok with you.


For startes the function `filterPrivateData` should take the _array_ of  employee records. So you can call the parameter `records`. This does not make a functional difference, but might help understand what is happening.

File: 1-JavaScript/Week3/homework/ex7-mindPrivacy.js [hunk 9]
```diff
@@ -30,8 +30,13 @@ const employeeRecords = [
...
   // TODO complete this function
+  
+  const {name, occupation, email} = employeeRecords[0];
+  const {gender, salary, ...rest } = employeeRecords[1];
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week3/homework/ex7-mindPrivacy.js:37]: You deconstructed in two different ways here, both are ok. But the idea is to loop over the entire array. So you could do that (among other ways) with a for loop:
```javascript
for (let record in records) {
    // do something with the record
} 
```

File: 1-JavaScript/Week3/homework/ex7-mindPrivacy.js [hunk 10]
```diff
@@ -30,8 +30,13 @@ const employeeRecords = [
...
+  const filterPrivate = [{name, occupation, email},rest];
```
Comment (bnhoogendijk) [inline 1-JavaScript/Week3/homework/ex7-mindPrivacy.js:38]: So this is where you managed to return a _new_ array in stead of the original. You could do that also by creating an empty array in the beginning of the function and push items to that new array within the loop.

**Review by bnhoogendijk — APPROVED** [review]

**Comment by bnhoogendijk** [comment]
Hi Amr, I think you did great on this weeks homework. In fact I approved your PR. Now I know you had some trouble, and it is not perfect, but you can be proud of what you achieved here! I am sure that over time you will keep improving. I left you some comments of which I hope they will help you. Keep up the good work!

File: 1-JavaScript/Week3/homework/ex3-tellFortune.js (hunk 3 above)
Comment (Amr-Yehia) [reply 1-JavaScript/Week3/homework/ex3-tellFortune.js:31]: In ```giveCompliments```  the array was separate so it was easy to call it with length property but here the array in function so I couldn't call the array! 
There is more than one array, which one I should use with .length!

I tried to call the main function inside the selectRandomly function and use one of the arrays bit it did not work, becaus of that i used that way in the home work
```js
function selectRandomly(arrays) {
  // TODO complete this function
  main();
  const random =
    numKids[Math.floor(Math.random() * numKids.length)];
  //const random = Math.floor(Math.random() * numKids[.]);
  const randomly = arrays[random];
  return randomly;
}
```

File: 1-JavaScript/Week3/homework/ex4-shoppingCart.js (hunk 5 above)
Comment (Amr-Yehia) [reply 1-JavaScript/Week3/homework/ex4-shoppingCart.js:29]: yes, I didn't make any difference.

File: 1-JavaScript/Week3/homework/ex7-mindPrivacy.js (hunk 9 above)
Comment (Amr-Yehia) [reply 1-JavaScript/Week3/homework/ex7-mindPrivacy.js:37]: I tried using if statement or for of loop inside for in loop unfortunately, I stauked ;(

**Pull request closed without merging**